### DISCUSSION TITLE
Callback Endpoint for workflow environments

### DISCUSSION THREAD

**Discussion by 6543** [woodpecker-ci/woodpecker#2954] [Request a Feature]
as noted in https://github.com/woodpecker-ci/woodpecker/pull/2935#issuecomment-1856590466
https://github.com/woodpecker-ci/woodpecker/discussions/2291#discussioncomment-7806713
and in lot of other issue/pull comments.

we need an other way beside the stdout to communicate with the agent to send/recive optional information:
- create optional UI
- expose in/outputs
- get additional infos

so I propose to expose an api via some socket (unix socet or udp/tcp ...)

And indicate that via some env var

a brain dump can be found here: https://github.com/woodpecker-ci/woodpecker/pull/2953

**Comment by lafriks** [comment]
I like the idea, probably Rest API to be more easily usable even from curl etc?

This could be potentially used also to upload pipeline files as build artifacts

**Reply by 6543** [reply]
well as for **Unix Sockets** [curl can do so too](https://kelcecil.com/curl/2015/12/07/query-unix-socket-with-curl.html) :D
so I would let it be a backend specific dessision if it's  a tcp or unix socket

**Reply by 6543** [reply]
-> for docker we can easily create a unix-socket and mount it via volume option into a container, so we dont have to worry about the network setup.
-> for local backend a tcp server might be best (works on windows too) and workflow-network == host-network (localhost)

**Reply by lafriks** [reply]
But that would not work if the pipeline is run on a different server than woodpecker server 🤔

**Reply by 6543** [reply]
well that's the whole point ... I proposed that the agent is doing that server and forward it to the server via grpc

as you could have setups where the agent can only reach the server via grpc (is an edgecase but doing so streamlines the responsibility)
and for other optional features the token is proposed to be also valid for the server api

**Reply by zc-devs** [reply]
How the socket supposed to be working on Kubernetes?

I'm lost, what connection you are trying to achieve?
Container/step -> Server? There are REST and gRPC already. I vote for REST as it's simpler for client: "Rest API to be more easily usable even from curl".
Container -> Agent? There is gRPC already.
